[21.0] RenderType changes

ApexModder opened the pull request:
This PR marks all commonly used `RenderType`/`RenderStateShard` fields/classes/methods as public, allowing modders to make custom render types easier, without needing to AT themselves or extend `RenderType` to access the `protected static` fields

With this change `NeoForgeRenderTypes.Internal` has been updated to no longer extend `RenderType`, this is to validate that the AT has been applied and continues to be applied in future updates.

ApexModder committed:
Mark all commonly used fields/methods public

ApexModder committed:
Stop extending RenderType to validate AT applied

embeddedt commented:
I'm not against this change in principle but the AT looks like an absolute chore to maintain.

Technici4n commented:
The ATs should be autogenerated.

ApexModder commented:
Rebased onto latest 20.6 changes

ApexModder commented:
_Replying to: Technici4n_
> The ATs should be autogenerated.

Unsure how you would go about having the ATs be autogenerated?

marchermans commented on src/main/resources/META-INF/accesstransformer.cfg:
```diff
@@ -121,26 +121,134 @@ public net.minecraft.client.particle.ParticleEngine$SpriteParticleRegistration
-protected-f net.minecraft.client.renderer.RenderStateShard$TextureStateShard mipmap # mipmap
 public net.minecraft.client.renderer.RenderStateShard$TexturingStateShard
 public net.minecraft.client.renderer.RenderStateShard$TransparencyStateShard
 public net.minecraft.client.renderer.RenderStateShard$WriteMaskStateShard
```
You do know that AT supports * operations, for fields and methods? Specifically for this reason.

ApexModder replied:
Weird wildcards worked just fine here, I remember them not working for me in the past which is why I didn't use them here initially

ApexModder committed:
Update to use wildcard ATs

ApexModder committed:
Add back missing setupState AT

ApexModder committed:
Port to 21.0

RenderStateShard.java.patch included to workaround issue with ATs

Cant AT `RenderStateShared.setupState` to none final as it is also modified via `RenderStateShard *` group

XFactHD commented on patches/net/minecraft/client/renderer/RenderStateShard.java.patch:
```diff
@@ -0,0 +1,11 @@
+     public static final double MAX_ENCHANTMENT_GLINT_SPEED_MILLIS = 8.0;
+     public final String name;
+-    public final Runnable setupState;
++    protected Runnable setupState; // Neo/AT-BUG: 'protected-f' does not work as field is also made public via AT field group
```
I don't like this (irrespective of the AT/JST bug). Patching the visibility back down prevents anyone else from using an AT on it and if we're making all fields public, then this field should not be non-final. I personally would prefer patching a second constructor into `TextureStateShard` which allows providing the runnables directly instead of having to override them afterwards. The patch could look something like this:
```patch
         public TextureStateShard(ResourceLocation pTexture, boolean pBlur, boolean pMipmap) {
-            super(() -> {
+            this(() -> {
                 TextureManager texturemanager = Minecraft.getInstance().getTextureManager();
                 texturemanager.getTexture(pTexture).setFilter(pBlur, pMipmap);
                 RenderSystem.setShaderTexture(0, pTexture);
             }, () -> {
-            });
+            }, pTexture, pBlur, pMipmap);
+        }
+        protected TextureStateShard(Runnable setupState, Runnable clearState, ResourceLocation pTexture, boolean pBlur, boolean pMipmap) {
+            super(setupState, clearState);
             this.texture = Optional.of(pTexture);
             this.blur = pBlur;
             this.mipmap = pMipmap;
         }
```

ApexModder replied:
While that would work in most cases, Neos `CustomizableTextureState` in `NeoForgeRenderTypes` requires the `setupState` to be non-final, as the modification made references `this`, which would not be possible via a `super(() -> { }, ...` since you cant reference `this` before `super`

https://github.com/neoforged/NeoForge/blob/c9c4d982b27128b390f6a6086426b2bf34cfc3e9/src/main/java/net/neoforged/neoforge/client/NeoForgeRenderTypes.java#L343-L353

notice the `this.` within the runnable, that would not be possible if passed via `super`.
`setupState` needs to be made non-final which due to the AT bug has to be done via a patch currently, unless I am mistaken that is?

XFactHD replied:
Dammit, yeah, you're right, it doesn't allow that. Another option that wouldn't even require a patch is to override `setupRenderState()` and perform the setup there, leaving the original setup runnable unused.

ApexModder committed:
Implement workaround for JST/AT issue

XFactHD commented on src/main/java/net/neoforged/neoforge/client/NeoForgeRenderTypes.java:
```diff
@@ -158,198 +155,203 @@ public RenderType get() {
+        private final Supplier<Boolean> blurSupplier;
+        private final Supplier<Boolean> mipmapSupplier;
+
         private CustomizableTextureState(ResourceLocation resLoc, Supplier<Boolean> blur, Supplier<Boolean> mipmap) {
```
More of a nitpick, but if we're touching all this anyway, then we might as well change this to use a `BooleanSupplier` instead of throwing around boxed `Boolean`s

ApexModder replied:
I was actually wondering the same thing tbh but thought better to leave as original implementation had them, but I can quickly swap them out

XFactHD commented on patches/net/minecraft/client/renderer/RenderStateShard.java.patch:
```diff
@@ -0,0 +1,11 @@
+     public static final double MAX_ENCHANTMENT_GLINT_SPEED_MILLIS = 8.0;
+     public final String name;
+-    public final Runnable setupState;
++    public Runnable setupState;
```
Please add the `final` back to remove this patch

XFactHD commented on src/main/java/net/neoforged/neoforge/client/NeoForgeRenderTypes.java:
```diff
@@ -158,198 +155,203 @@ public RenderType get() {
+
+        @Override
+        public void setupRenderState() {
+            // must be done before super call
+            // as super uses the `blur` and `mipmap` fields
+            // within the `setupState` runnable
```
This comment can be a single line

ApexModder committed:
Remove RenderStateShard patch

ApexModder committed:
Single line comment in NeoForgeRenderTypes

ApexModder committed:
Swap out `Supplier<Boolean>` for `BooleanSupplier`

XFactHD committed:
Merge branch '1.21.x' into pr/render-type
